### PR TITLE
Migrate LEAN to .NET Core 3.0

### PR DESCRIPTION
**Pull request by gsalaz98**
#### Description

Upgrades LEAN to .NET Core 3.0 (we can lower the required version to 2.0+ easily). 

Disabled Views and Visual Studio plugin project to have project be compatible with .NET Core.

NOTE: Python algorithms will not work without an update to pythonnet.
#### Related Issue

#452 
#### Motivation and Context

Performance gains on Mac OSX and Linux (100% on BasicTemplateAlgorithm and 50% on Tick data on Mac OSX)
#### Requires Documentation Change

Build process has not changed, but Linux + Mac OSX building is now more easy via the dotnet command line utility. Mono would not be required anymore.
#### How Has This Been Tested?

Able to run algorithms locally. I haven't ran regression tests and not all unit tests. I suspect a lot of work is still needed to have this equal to LEAN master, but it's a start
#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes.  Unable to run unit tests locally. Did run some tests on the Composer that completed successfully because I altered it heavily.
- [ ] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Brokerages/QuantConnect.Brokerages.csproj [hunk 1]
```diff
@@ -678,13 +213,20 @@
     </None>
   </ItemGroup>
   <ItemGroup />
-  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <ItemGroup>
+    <PackageReference Include="IKVM" Version="8.1.5717" />
```
Comment (gsalaz98) [inline Brokerages/QuantConnect.Brokerages.csproj]: Update IKVM

File: Common/Data/UniverseSelection/UserDefinedUniverse.cs [hunk 2]
```diff
@@ -17,9 +17,9 @@
 using System.Collections.Generic;
 using System.Collections.Specialized;
 using System.Linq;
+using QuantConnect.Util;
```
Comment (gsalaz98) [inline Common/Data/UniverseSelection/UserDefinedUniverse.cs]: Fixup

File: Common/Global.cs [hunk 3]
```diff
@@ -365,7 +365,8 @@ public enum DataFeedEndpoint
         /// Getting datafeed from a QC-Live-Cloud
         LiveTrading,
         /// Database
-        Database
+        Database,
+        Tradier
```
Comment (gsalaz98) [inline Common/Global.cs]: Remove tradier

File: Common/OS.cs [hunk 4]
```diff
@@ -123,23 +122,22 @@ public static decimal CpuUsage
...
-                            "% Processor Time",
-                            IsWindows ? Process.GetCurrentProcess().ProcessName : Process.GetCurrentProcess().Id.ToString());
-                    }
-                    catch (Exception exception)
-                    {
-                        Log.Error(exception);
-                        return 0;
-                    }
-                }
+                var startTime = DateTime.UtcNow;
```
Comment (gsalaz98) [inline Common/OS.cs]: Rewrite this

File: Common/Util/Composer.cs [hunk 5]
```diff
@@ -270,7 +268,7 @@ public IEnumerable<T> GetExportedValues<T>()
                 {
                     _composableParts.Wait();
                 }
-                values = _compositionContainer.GetExportedValues<T>().ToList();
+                values = _compositionContainer.GetExports<T>().ToList();
```
Comment (gsalaz98) [inline Common/Util/Composer.cs]: Make demo API and test

File: Engine/DataFeeds/ZipDataCacheProvider.cs [hunk 6]
```diff
@@ -57,6 +58,9 @@ public ZipDataCacheProvider(IDataProvider dataProvider, bool isDataEphemeral = t
         /// </summary>
         public Stream Fetch(string key)
         {
+            // Will throw while trying to read ZIP file if this line is not included
+            Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
```
Comment (gsalaz98) [inline Engine/DataFeeds/ZipDataCacheProvider.cs]: link to stackoverflow issue

File: Engine/Results/BacktestingResultHandler.cs [hunk 7]
```diff
@@ -31,12 +31,14 @@
...
 using QuantConnect.Lean.Engine.Alphas;
 using QuantConnect.Securities;
+using System.Composition;
 
 namespace QuantConnect.Lean.Engine.Results
 {
     /// <summary>
     /// Backtesting result handler passes messages back from the Lean to the User.
     /// </summary>
+    [Export(typeof(IResultHandler))]
```
Comment (gsalaz98) [inline Engine/Results/BacktestingResultHandler.cs]: remove

File: Engine/Results/DesktopResultHandler.cs [hunk 8]
```diff
@@ -28,12 +28,14 @@
...
 using System.Diagnostics;
 using QuantConnect.Securities;
+using System.Composition;
 
 namespace QuantConnect.Lean.Engine.Results
 {
     /// <summary>
     /// Desktop Result Handler - Desktop GUI Result Handler for Piping Results to WinForms:
     /// </summary>
+    [Export(typeof(IResultHandler))]
```
Comment (gsalaz98) [inline Engine/Results/DesktopResultHandler.cs]: remove

File: Engine/Results/LiveTradingResultHandler.cs [hunk 9]
```diff
@@ -40,6 +41,7 @@ namespace QuantConnect.Lean.Engine.Results
     /// Live trading result handler implementation passes the messages to the QC live trading interface.
     /// </summary>
     /// <remarks>Live trading result handler is quite busy. It sends constant price updates, equity updates and order/holdings updates.</remarks>
+    [Export(typeof(IResultHandler))]
```
Comment (gsalaz98) [inline Engine/Results/LiveTradingResultHandler.cs]: remove

File: Engine/Results/RegressionResultHandler.cs [hunk 10]
```diff
@@ -26,6 +27,7 @@ namespace QuantConnect.Lean.Engine.Results
     /// Provides a wrapper over the <see cref="BacktestingResultHandler"/> that logs all order events
     /// to a separate file
     /// </summary>
+    [Export(typeof(IResultHandler))]
```
Comment (gsalaz98) [inline Engine/Results/RegressionResultHandler.cs]: remove

File: Launcher/Program.cs [hunk 11]
```diff
@@ -79,7 +76,7 @@ static void Main(string[] args)
             {
                 leanEngineSystemHandlers = LeanEngineSystemHandlers.FromConfiguration(Composer.Instance);
             }
-            catch (CompositionException compositionException)
+            catch (Exception compositionException)
```
Comment (gsalaz98) [inline Launcher/Program.cs]: Review exception catching

File: Launcher/QuantConnect.Lean.Launcher.csproj [hunk 12]
```diff
@@ -232,13 +226,14 @@
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
   </ItemGroup>
-  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
-  <Import Project="..\packages\Baseclass.Contrib.Nuget.Output.2.2.0-xbuild02\build\net40\Baseclass.Contrib.Nuget.Output.targets" Condition="Exists('..\packages\Baseclass.Contrib.Nuget.Output.2.2.0-xbuild02\build\net40\Baseclass.Contrib.Nuget.Output.targets')" />
+  <ItemGroup>
+    <PackageReference Include="DynamicInterop" Version="0.7.4" />
```
Comment (gsalaz98) [inline Launcher/QuantConnect.Lean.Launcher.csproj]: flagged for review

File: Logging/ILogHandler.cs [hunk 13]
```diff
@@ -14,14 +14,14 @@
...
-using System.ComponentModel.Composition;
+using System.Composition.Hosting;
 
 namespace QuantConnect.Logging
 {
     /// <summary>
     /// Interface for redirecting log output
     /// </summary>
-    [InheritedExport(typeof(ILogHandler))]
+    //[InheritedExport(typeof(ILogHandler))]
```
Comment (gsalaz98) [inline Logging/ILogHandler.cs]: remove comment

File: QuantConnect.Lean.sln.DotSettings [hunk 14]
```diff
@@ -7,6 +7,7 @@
 	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=PrivateInstanceFields/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="_" Suffix="" Style="aaBb" /&gt;</s:String>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ECSharpKeepExistingMigration/@EntryIndexedValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ECSharpPlaceEmbeddedOnSameLineMigration/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ECSharpUseContinuousIndentInsideBracesMigration/@EntryIndexedValue">True</s:Boolean>
```
Comment (gsalaz98) [inline QuantConnect.Lean.sln.DotSettings]: remove

File: Tests/QuantConnect.Tests.csproj [hunk 15]
```diff
@@ -810,10 +416,6 @@
       <Project>{73fb2522-c3ed-4e47-8e3d-afad48a6b888}</Project>
       <Name>QuantConnect.Indicators</Name>
     </ProjectReference>
-    <ProjectReference Include="..\Jupyter\QuantConnect.Jupyter.csproj">
-      <Project>{9561d14a-467e-40ad-928e-ee9f758d7d98}</Project>
```
Comment (gsalaz98) [inline Tests/QuantConnect.Tests.csproj]: review further

File: Tests/RegressionAlgorithms/Test_OnEndOfDay.cs [hunk 16]
```diff
@@ -36,7 +38,7 @@ public override void Initialize()
             AddSecurity(SecurityType.Equity, symbol, Resolution.Minute);
         }
 
-        public override void OnTradeBar(Dictionary<string, TradeBar> data)
+        public void OnTradeBar(Dictionary<string, TradeBar> data)
```
Comment (gsalaz98) [inline Tests/RegressionAlgorithms/Test_OnEndOfDay.cs]: further review

File: ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs [hunk 17]
```diff
@@ -357,7 +357,7 @@ private static bool DecompressOpraFile(FileInfo compressedRawDatafile, FileInfo
             {
                 try
                 {
-                    BZip2.Decompress(fileToDecompressAsStream, decompressedStream);
+                    BZip2.Decompress(fileToDecompressAsStream, decompressedStream, true);
```
Comment (gsalaz98) [inline ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs]: review further

File: UserInterface/WinForms/LeanWinForm.Designer.cs [hunk 18]
```diff
@@ -11,7 +11,7 @@ partial class LeanWinForm
         /// Clean up any resources being used.
         /// </summary>
         /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
-        protected override void Dispose(bool disposing)
+        protected void Dispose(bool disposing)
```
Comment (gsalaz98) [inline UserInterface/WinForms/LeanWinForm.Designer.cs]: Delete UserInterface project

Comment (gsalaz98) [reply UserInterface/WinForms/LeanWinForm.Designer.cs]: And visual studio plugin

File: Engine/DataFeeds/ZipDataCacheProvider.cs (hunk 6 above)
Comment (mchandschuh) [reply Engine/DataFeeds/ZipDataCacheProvider.cs]: Might also want to move this registration to a static constructor or somewhere else that ensure we only run it once per `AppDomain`

**Comment by mchandschuh** [comment]
This looks like it's still a work in progress, so I'll save the review until it reaches a slightly more stable state.

That being said, I think the move to .net core is a great one. Where possible we should strive to make the libraries compile against the .net standard libraries as they have the best portability and the most reusability (they can be referenced from .net framework projects as well).

**Comment by gsalaz98** [comment]
Closing for now; will continue development on https://github.com/gsalaz98/Lean/tree/feature-452-dotnet-core-test 

Lots of work has yet to be done. Most notably, porting unit tests to NUnit 3, testing composer changes, pythonnet support (python support in general)

**Pull request closed without merging**